ext/crates/fp/src/limb.rs

```diff
@@ -11,6 +10,148 @@ pub(crate) struct LimbBitIndexPair {
+    /// tradeoff is potentially worse.
+    limbs: usize,
+}
+
+impl<const P: u32> LimbLength<P> {
+    pub const fn from_limbs(limbs: usize) -> Self {
+        let logical = entries_per_limb_const::<P>() * limbs;
+        Self {
+            start: 0,
+            end: logical,
+            limbs,
+        }
+    }
+
+    pub const fn from_logical(logical: usize) -> Self {
+        let limbs = number::<P>(logical);
+        Self {
+            start: 0,
+            end: logical,
+            limbs,
+        }
+    }
+
+    /// Returns a `LimbLength` describing a vector starting at entry `start` and ending at entry
+    /// `end`. Most methods assume that `self.start < entries_per_limb_const::<P>()`, so it advised
+    /// to call `apply_shift` before using the return value.
+    pub fn from_start_end(start: usize, end: usize) -> Self {
+        let limb_range = range::<P>(start, end);
+        Self {
+            start,
+            end,
+            limbs: limb_range.end - limb_range.start,
+        }
+    }
+
+    #[inline]
+    pub const fn limbs(&self) -> usize {
+        self.limbs
+    }
+
+    #[inline]
+    pub const fn logical(&self) -> usize {
+        self.end - self.start
+    }
+
+    pub const fn contains(&self, other: &Self) -> bool {
+        self.start <= other.start && self.end >= other.end
+    }
+
+    /// Shift the entire `LimbLength` backwards so that the start of the range belongs to the first
+    /// limb, and return the number of limbs shifted.
+    pub fn apply_shift(&mut self) -> usize {
+        let entries_per = entries_per_limb_const::<P>();
+        let offset = self.start / entries_per;
+        self.start -= offset * entries_per;
+        self.end -= offset * entries_per;
+        offset
+    }
+
+    pub fn restrict_to(&self, other: Self) -> Self {
+        debug_assert!(self.start + other.end <= self.end);
+        Self::from_start_end(other.start + self.start, other.end + self.start)
+    }
+
+    pub fn trim_start(&self, offset: usize) -> Self {
+        debug_assert_eq!(self.start, 0);
+        assert_eq!(offset % entries_per_limb_const::<P>(), 0);
+        let limb_shift = offset / entries_per_limb_const::<P>();
+        Self {
+            start: self.start,
+            end: self.end - offset,
+            limbs: self.limbs - limb_shift,
+        }
+    }
+
+    #[inline]
+    pub fn bit_offset(&self) -> usize {
+        self.start * bit_length_const::<P>()
+    }
+
+    #[inline]
+    pub fn limb_range(&self) -> Range<usize> {
+        range::<P>(self.start, self.end)
+    }
+
+    /// This function underflows if `self.start + self.logical() == 0`, which happens if and only if
+    /// we are taking a slice of width 0 at the start of an `FpVector`. This should be a very rare
+    /// edge case. Dealing with the underflow properly would probably require using `saturating_sub`
+    /// or something of that nature, and that has a nontrivial (10%) performance hit.
```

Comment:
Can we insert a debug_assert that this doesn't happen? I assume it would be a programmer error if they allow this to happen. And the docs can simply say that the function cannot be called on such a range.

Fun fact: For a long time, `FpVector::<2>::add` was incorrectly implemented to ignore the coefficient, since we often check if the coefficient is 0 before we call `add`. It was discovered when it lead to a very confusing bug.

Reply:
I made the documentation hopefully a bit more explicit